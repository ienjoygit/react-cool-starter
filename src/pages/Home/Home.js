import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Helmet from 'react-helmet';

import { usersAction } from '../../actions';
import { UserList } from '../../components';
import styles from './styles.scss';

// Export this for unit testing more easily
export class Home extends PureComponent {
  componentDidMount() {
    const { fetchUsersIfNeeded } = this.props;

    fetchUsersIfNeeded();
  }

  renderUserList = () => {
    const { readyStatus, list } = this.props;

    if (
      !readyStatus ||
      readyStatus === 'USERS_INVALID' ||
      readyStatus === 'USERS_REQUESTING'
    )
      return <p>Loading...</p>;

    if (readyStatus === 'USERS_FAILURE')
      return <p>Oops, Failed to load list!</p>;

    return <UserList list={list} />;
  };

  render() {
    return (
      <div className={styles.Home}>
        <Helmet title="Home" />
        {this.renderUserList()}
      </div>
    );
  }
}

Home.propTypes = {
  readyStatus: PropTypes.string.isRequired,
  list: PropTypes.array.isRequired,
  fetchUsersIfNeeded: PropTypes.func.isRequired
};

const mapStateToProps = ({ home: { readyStatus, list } }) => ({
  readyStatus,
  list
});

const mapDispatchToProps = dispatch => ({
  fetchUsersIfNeeded: () => dispatch(usersAction.fetchUsersIfNeeded())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Home);
