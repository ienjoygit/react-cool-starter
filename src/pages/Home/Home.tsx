import React, { useEffect, memo } from "react";
import { useDispatch, useSelector, shallowEqual } from "react-redux";
import { Helmet } from "react-helmet";

import { usersAction } from "../../actions";
import { UserList } from "../../components";
import { AppState } from "../../types";
import styles from "./styles.scss";

const Home = (): JSX.Element => {
  const dispatch = useDispatch();
  const { readyStatus, list } = useSelector(
    ({ home }: AppState) => home,
    shallowEqual
  );

  useEffect(() => {
    dispatch(usersAction.fetchUsersIfNeeded());
  }, []);

  const renderUserList = () => {
    if (!readyStatus || readyStatus === "invalid" || readyStatus === "request")
      return <p>Loading...</p>;

    if (readyStatus === "failure") return <p>Oops, Failed to load list!</p>;

    return <UserList list={list} />;
  };

  return (
    <div className={styles.Home}>
      <Helmet title="Home" />
      {renderUserList()}
    </div>
  );
};

export default memo(Home);
