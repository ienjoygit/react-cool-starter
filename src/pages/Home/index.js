/* @flow */
import React from 'react';
import loadable from '@loadable/component';
import { Loading, ErrorBoundary } from '../../components';

const Home = loadable(() => import('./Home'), {
  fallback: Loading
});

export default () => (
  <ErrorBoundary>
    <Home />
  </ErrorBoundary>
);
