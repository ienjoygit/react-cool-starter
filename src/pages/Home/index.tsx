import loadable from "@loadable/component";

import { Loading, ErrorBoundary } from "../../components";
import { loadData } from "./Home";

const Home = loadable(() => import("./Home"), {
  fallback: <Loading />,
});

export default (): JSX.Element => (
  <ErrorBoundary>
    <Home />
  </ErrorBoundary>
);
export { loadData };
