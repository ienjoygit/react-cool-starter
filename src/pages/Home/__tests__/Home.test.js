import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import { Home } from '../Home';

describe('<Home />', () => {
  const tree = (props, actions) =>
    renderer
      .create(
        <MemoryRouter>
          <Home {...props} {...actions} />
        </MemoryRouter>
      )
      .toJSON();

  it('should call fetchUsersIfNeeded when componentDidMount', () => {
    const mockAction = jest.fn();
    const props = {
      home: {}
    };
    const actions = {
      fetchUsersIfNeeded: mockAction
    };

    mount(
      <MemoryRouter>
        <Home {...props} {...actions} />
      </MemoryRouter>
    );

    expect(mockAction).toHaveBeenCalled();
  });

  it('renders the loading status if data invalid', () => {
    const props = { home: { readyStatus: 'invalid' } };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  it('renders the loading status if requesting data', () => {
    const props = { home: { readyStatus: 'request' } };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  it('renders an error if loading failed', () => {
    const props = { home: { readyStatus: 'failure' } };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  it('renders the <UserList /> if loading was successful', () => {
    const props = {
      home: {
        readyStatus: 'success',
        list: [{ id: '1', name: 'Welly' }]
      }
    };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });
});
