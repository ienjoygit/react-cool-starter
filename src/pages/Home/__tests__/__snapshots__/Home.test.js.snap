// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders an error if loading failed 1`] = `
<div
  className="Home"
>
  <p>
    Loading...
  </p>
</div>
`;

exports[`<Home /> renders the <UserList /> if loading was successful 1`] = `
<div
  className="Home"
>
  <p>
    Loading...
  </p>
</div>
`;

exports[`<Home /> renders the loading status if data invalid 1`] = `
<div
  className="Home"
>
  <p>
    Loading...
  </p>
</div>
`;

exports[`<Home /> renders the loading status if requesting data 1`] = `
<div
  className="Home"
>
  <p>
    Loading...
  </p>
</div>
`;
